File: src/leetcode/java/BinaryTreeMaximumPathSum_124.java
package leetcode.java.hard;

public class BinaryTreeMaximumPathSum_124 {
  private int max;

  public int maxPathSum(TreeNode root) {
    max = Integer.MIN_VALUE;
    caculateMax(root);
    return max;
  }

  private int caculateMax(TreeNode node) {
    if (node == null) return 0;

    int lMax = Math.max(0, caculateMax(node.left));
    int rMax = Math.max(0, caculateMax(node.right));
    max = Math.max(max, lMax + rMax + node.val);
    return Math.max(lMax, rMax) + node.val;
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}

File: src/leetcode/python/BinaryTreeMaximumPathSum_124.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class BinaryTreeMaximumPathSum_124(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def calculate(node):
            if not node:
                return 0
            lmax = max(0, calculate(node.left))
            rmax = max(0, calculate(node.right))
            path_max[0] = max(path_max[0], lmax + rmax + node.val)
            return max(lmax, rmax) + node.val

        path_max = [0]
        if root:
            path_max[0] = root.val
        else:
            return 0
        calculate(root)
        return path_max[0]
